src/main/scala/is/hail/io/index/IndexWriter.scala

```diff
@@ -0,0 +1,165 @@
+package is.hail.io.index
+
+import is.hail.expr.types._
+import is.hail.io.CodecSpec
+import is.hail.rvd.RVDContext
+import is.hail.utils._
+import is.hail.utils.richUtils.ByteTrackingOutputStream
+import org.apache.hadoop.conf.Configuration
+import org.json4s.Formats
+import org.json4s.jackson.Serialization
+
+case class IndexMetadata(
+  fileVersion: Int,
+  branchingFactor: Int,
+  keyType: String,
+  nKeys: Long,
+  indexPath: String,
+  rootOffset: Long,
+  attributes: Map[String, Any]
+)
+
+class IndexWriter(
+  conf: Configuration,
+  path: String,
+  keyType: Type,
+  branchingFactor: Int = 1024,
+  attributes: Map[String, Any] = Map.empty[String, Any]) {
+
+  private var elementIdx = 0L
+  private var rootOffset = 0L
+
+  private val ctx = RVDContext.default
+  private val rvb = ctx.rvb
+  private val region = ctx.freshRegion
+  rvb.set(region)
+
+  private val leafNode = new LeafNodeBuilder(keyType)
+  private val internalNodes = new ArrayBuilder[InternalNodeBuilder]()
+
+  private val trackedOS = new ByteTrackingOutputStream(conf.unsafeWriter(path + "/index"))
+  private val codecSpec = CodecSpec.default
+  private val leafEncoder = codecSpec.buildEncoder(leafNode.typ)(trackedOS)
+  private val internalEncoder = codecSpec.buildEncoder(InternalNodeBuilder.typ(keyType))(trackedOS)
+
+  private def writeInternalNode(node: InternalNodeBuilder): (Long, Long, Any, Long) = {
+    val fileOffset = trackedOS.bytesWritten
+    rootOffset = fileOffset
+
+    val firstKey = node.firstKey
+    val firstKeyOffset = node.firstOffset
+    val firstIndex = node.firstIndex
+
+    internalEncoder.writeByte(1)
+
+    val regionOffset = node.write(rvb)
+    internalEncoder.writeRegionValue(region, regionOffset)
+    internalEncoder.flush()
+
+    region.clear()
+    node.clear()
+
+    (fileOffset, firstIndex, firstKey, firstKeyOffset)
```

Comment:
I feel you pack and unpack these in a few places.  Would be simpler to make a case class for this?  Or just reuse Internal/LeafChild?